### ISSUE TITLE
Watching on Linux fails with `Received event for unknown watch descriptor 527`

### ISSUE THREAD

**Issue by wolfs** [gradle/gradle#13079]
When running the VFS retention integration tests without isolated daemons on Linux (#12604), some recorded an exception in file watching ([Build Scan](https://e.grdev.net/s/p2jtphe6uhpho/tests/:testingJvm:vfsRetentionIntegTest/org.gradle.testing.TestTaskIntegrationTest/re-runs%20tests%20when%20resources%20are%20renamed%20%5B4.13%5D#1)):

```
Error while receiving file changes
  net.rubygrapefruit.platform.NativeException: Received event for unknown watch descriptor 527
    at net.rubygrapefruit.platform.internal.jni.AbstractFileEventFunctions$NativeFileWatcher.executeRunLoop0(Native Method)
    at net.rubygrapefruit.platform.internal.jni.AbstractFileEventFunctions$NativeFileWatcher.access$100(AbstractFileEventFunctions.java:156)
    at net.rubygrapefruit.platform.internal.jni.AbstractFileEventFunctions$NativeFileWatcher$1.run(AbstractFileEventFunctions.java:169)
```

The tests which fail seem to be renaming files, which means we get the renamed event on Linux which we currently don't handle and causes the watching to be stopped.

Re-running the build did not reproduce the problem :(

---
cc: @gradle/build-cache

**Comment by wolfs** [comment]
Looks like we are getting an `IN_DELETE_SELF` (`0x400`) event after the `IN_IGNORED` (`0x8000`) event. Since the `IN_IGNORED` event removes the watch descriptor from `recentlyRemovedWatchPoints`

```
[NativeLogger] Watching path /.../build/reports/tests with watch descriptor 1299

....

[NativeLogger] Processed 21 events
[NativeLogger] Cancelling watching path /.../build/reports/tests
[NativeLogger] Cancelled watching path /.../build/reports/tests
[NativeLogger] Processing 1504 bytes worth of events

...


[NativeLogger] Finished watching '/.../build/reports/tests/test'
[NativeLogger] Event mask: 0x40000200 for test (wd = 1299, cookie = 0x0, len = 16
)
[NativeLogger] Ignoring incoming events for recently removed watch descriptor 1299
[NativeLogger] Event mask: 0x8000 for  (wd = 1299, cookie = 0x0, len = 0)
[NativeLogger] Event mask: 0x400 for  (wd = 1299, cookie = 0x0, len = 0)
```

**Comment by wolfs** [comment]
We also seem to be receiving strange events with watch descriptor `-1`:

```
Finished watching '/.../build/generated/sources/headers/java/test'
Path is not watched: /.../build/generated/sources/annotationProcessor/java/test
Path is not watched: /.../build/generated/sources/annotationProcessor/java
Cancelling watching path /.../build/generated/sources/headers/java
Cancelled watching path /.../build/generated/sources/headers/java
Path is not watched: /.../build/generated/sources/headers/java/test
Event mask: 0x40000200 for test (wd = -1, cookie = 0x0, len = 16)
Error while receiving file changes
net.rubygrapefruit.platform.NativeException: Received event for unknown watch descriptor -1 for event 1073742336
        at net.rubygrapefruit.platform.internal.jni.AbstractFileEventFunctions$NativeFileWatcher.executeRunLoop0(Native Method)
        at net.rubygrapefruit.platform.internal.jni.AbstractFileEventFunctions$NativeFileWatcher.access$100(AbstractFileEventFunctions.java:156)
        at net.rubygrapefruit.platform.internal.jni.AbstractFileEventFunctions$NativeFileWatcher$1.run(AbstractFileEventFunctions.java:169)
```

I wonder if we should ignore events with watch descriptor `-1`...

**Comment by lptr** [comment]
I'll take a look at supporting `IN_MOVE_SELF` tomorrow. I guess sending an `INVALIDATE` event for the path should be okay.

**Comment by wolfs** [comment]
I don't think handling `IN_MOVE_SELF` will help. There are two problems here:
- We receive events with watch descriptor `-1`.
- We receive events for watch descriptors after receiving the `IN_IGNORED` event.

Both things seem to happen while the directory for the test is deleted.

It seems to be enough to
- ignore events with watch descriptor `-1`.
- only remove the watch descriptors from `recentlyRemovedWatchPoints` for `IN_IGNORED` events after processing the current batch of events (https://github.com/gradle/native-platform/blob/173d2ae3977466a69144b245edc9d7b864d70faa/src/file-events/cpp/linux_fsnotifier.cpp#L146-L160).

I currently don't have another good solution for the problem, nor can I reproduce it in `native-platform` alone.

**Comment by lptr** [comment]
I cannot reproduce the issue with receiving events for an already ignored watch descriptor. :/

About `wd == -1`, it seems strange that `wd` is explicitly declared as a signed `int`, whereas the other values are explicitly unsigned (`inotify.h`):

```c
/* Structure describing an inotify event.  */
struct inotify_event
{
  int wd;		/* Watch descriptor.  */
  uint32_t mask;	/* Watch mask.  */
  uint32_t cookie;	/* Cookie to synchronize two events.  */
  uint32_t len;		/* Length (including NULs) of name.  */
  char name __flexarr;	/* Name.  */
};
```

**Comment by lptr** [comment]
>            `IN_Q_OVERFLOW`
                  Event queue overflowed (`wd` is -1 for this event).

Are we catching an overflow? http://man7.org/linux/man-pages/man7/inotify.7.html

**Comment by wolfs** [comment]
Doesn't seem to be an overflow:
```
Event mask: 0x40000200 for pkg (wd = -1, cookie = 0x0, len = 16)
```

**Comment by lptr** [comment]
Looks like we are hitting a bug. I'm just wondering if the cause is an overflow. One way this could happen is if the kernel noticed an overflow, and then replaced the `wd` of the first event with `-1`, but "forgot" to update the event mask. Yeah, it's far fetched, and not sure how it would change things having been proven correct. But I'm at my wit's end here, catching at straws...

**Comment by wolfs** [comment]
One of the test that fails is [`JUnitFilteringIntegrationTest.task is out of date when --tests argument changes`](https://github.com/gradle/gradle/blob/2f39c260e089b32c9ac378652da64a0aa0857f16/subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractTestFilteringIntegrationTest.groovy#L211-L232).

```
Processing 48 bytes worth of events
Event mask: 0x200 for TEST-FooTest.xml (wd = 1336, cookie = 0x0, len = 32)
Processed 1 events
Cancelling watching path '/home/wolf/gradle/subprojects/testing-jvm/build/tmp/test files/JUnitFilteringIntegrationTest/task_is_out_of_date...t_changes/5zxzo/build/test-results/test' (wd = 1336)
Processing 64 bytes worth of events
Event mask: 0x8000 for  (wd = 1336, cookie = 0x0, len = 0)
Finished watching (wd = 1336)
Event mask: 0x100 for TEST-FooTest.xml (wd = -1, cookie = 0x0, len = 32)
Processing 208 bytes worth of events
Error while receiving file changes
net.rubygrapefruit.platform.NativeException: Received event for unknown watch descriptor -1
        at net.rubygrapefruit.platform.internal.jni.AbstractFileEventFunctions$NativeFileWatcher.executeRunLoop0(Native Method)
        at net.rubygrapefruit.platform.internal.jni.AbstractFileEventFunctions$NativeFileWatcher.access$100(AbstractFileEventFunctions.java:156)
        at net.rubygrapefruit.platform.internal.jni.AbstractFileEventFunctions$NativeFileWatcher$1.run(AbstractFileEventFunctions.java:169)
Event mask: 0x200 for default-package.html (wd = 1340, cookie = 0x0, len = 32)
Event mask: 0x400 for  (wd = 1340, cookie = 0x0, len = 0)
Event mask: 0x8000 for  (wd = 1340, cookie = 0x0, len = 0)
```

The test uses `DefaultTestExecutionResult` to read the `TEST-FooTest.xml` file. This class uses [`XmlSlurper` to parse the xml file](https://github.com/gradle/gradle/blob/e38b4115714868523bd5e0a1ba190f35fa9a7001/subprojects/internal-testing/src/main/groovy/org/gradle/integtests/fixtures/JUnitXmlTestExecutionResult.groovy#L77).

The test task itself writes the xml output file [here](https://github.com/gradle/gradle/blob/7e461f7a72d56c04d94c436ccdb1ef184fa72f5b/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/junit/result/Binary2JUnitXmlReportGenerator.java#L113-L125).

**Comment by wolfs** [comment]
Fixed by gradle/native-platform#212.
